chore(gatsby): convert component-data-dependencies to typescript

Kornil opened the pull request:
Convert `src/redux/reducers/component-data-dependencies.ts` and test file to TypeScript.

**There is one change that can possibly affect production**
In `src/redux/reducers/component-data-dependencies.ts` the action `CREATE_COMPONENT_DEPENDENCY` was referencing a `payload.id` in line 19, 20, 31, 32. I found the key to not be used in the action itself, so I removed it. Let me know your opinion, I haven't seen it used.

## Related Issues

https://github.com/gatsbyjs/gatsby/issues/21995

pieh commented:
> **There is one change that can possibly affect production**
> In `src/redux/reducers/component-data-dependencies.ts` the action `CREATE_COMPONENT_DEPENDENCY` was referencing a `payload.id` in line 19, 20, 31, 32. I found the key to not be used in the action itself, so I removed it. Let me know your opinion, I haven't seen it used.

This should be fine - it seems like legacy code that originally was added in https://github.com/gatsbyjs/gatsby/pull/1503, but feature later on was removed as part of v2 release

pieh commented on packages/gatsby/src/redux/types.ts:
```diff
@@ -149,7 +149,7 @@ export interface IGatsbyState {
   }
   componentDataDependencies: {
-    nodes: Map<string, Set<string>>
+    nodes: Map<string | number, Set<string>>
```
This should be just `string` (keys should be node ids which are strings). 

What caused the need for ` | number`? If there is any code that use numbers we need to fix it there.

This also potentially cause some unnecessary casting happening right now in this PR

Kornil replied:
In the test file, some nodes are mocked as `number`, I'll go ahead and fix there too.

pieh replied:
I'll go ahead and commit a change because you adjusted tests but seem to missed this one :) I already verified locally that typecheck is happy after dropping ` | number` here so I should update it in few minutes

pieh commented on packages/gatsby/src/redux/reducers/__tests__/page-data-dependencies.ts:
```diff
@@ -95,11 +97,13 @@ describe(`add page data dependency`, () => {
-    expect(state.connections.get(`MarkdownRemark`).size).toEqual(2)
-    expect(state.nodes.get(1).size).toEqual(2)
+    expect(
+      (state.connections.get(`MarkdownRemark`) as Set<string>).size
```
If casting here is to avoid potential `undefined` let's use non-null assertion operator
```suggestion
      state.connections.get(`MarkdownRemark`)!.size
```

if it it's not about that - then why is this needed?

Similarly line below (+ question why is this `number | string` in first place - but there is other comment closer to typing source about this

pieh replied:
Actually - let's use optional chaining which is much safer and still will work for this use case
```suggestion
      state.connections.get(`MarkdownRemark`)?.size
```

Kornil commented:
Thank you for the tough review @pieh! Added all suggestions, and some.

I removed the `number` type from `nodeId` and fixed the tests to reflect that. Also removed type casting in tests for optional chaining and removed casting from reducer file for the preferred non-null operator.

pieh commented on packages/gatsby/src/redux/reducers/component-data-dependencies.ts:
```diff
@@ -12,36 +17,36 @@ module.exports = (
+        let existingPaths: Set<string> = new Set()
         if (state.connections.has(action.payload.connection)) {
-          existingPaths = state.connections.get(action.payload.connection)
+          existingPaths = state.connections.get(action.payload.connection)!
```
this is so weird that TS is not smart enough to know that this for sure exists given the check above it, but 🤷‍♂️
